### PR TITLE
Use MethodHandles to Run HttpClient Warmup

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR introduces using `MethodHandles` to run warmup on an `HttpClient` when the API exists. If the API doesn't exist it won't be attempted and the application will continue running as normal. This change should allow Cosmos to work if Reactor Netty resolves to a version lower than `1.0.3` in the `1.0.x` major-minor version set.

This is similar to the changes made in #19918 which allowed `azure-core` to work with both Jackson 2.11 and 2.12.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java [hunk 1]
```diff
@@ -77,11 +95,29 @@ public static ReactorNettyClient createWithConnectionProvider(ConnectionProvider
...
+    private static void attemptToWarmupHttpClient(ReactorNettyClient reactorNettyClient) {
+        // Warmup wasn't found, so don't attempt it.
+        if (HTTP_CLIENT_WARMUP == null) {
+            return;
+        }
+
+        try {
+            ((Mono<?>) HTTP_CLIENT_WARMUP.invoke(reactorNettyClient.httpClient)).block();
+        } catch (ClassCastException | WrongMethodTypeException throwable) {
+            // Invocation failed.
```
Comment (anuchandy) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:114]: Should we log this (i.e. something we want to investigate when log appear) or fine with ignoring these exceptions?

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:114]: I think we should, but not on info level, may be debug level logging.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java [hunk 2]
```diff
@@ -43,6 +47,21 @@
...
     private static final Logger logger = LoggerFactory.getLogger(ReactorNettyClient.class.getSimpleName());
 
+    private static final MethodHandle HTTP_CLIENT_WARMUP;
+
+    static {
+        MethodHandle httpClientWarmup = null;
+        try {
+            httpClientWarmup = MethodHandles.publicLookup()
+                .findVirtual(reactor.netty.http.client.HttpClient.class, "warmup", MethodType.methodType(Mono.class));
+        } catch (Throwable throwable) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:57]: Though this is nit, but catching Throwable here is catching everything. 
Shouldn't we just catch the exception which is thrown by `publicLookup` or `findVirtual` ?

Comment (alzimmermsft) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:57]: Good catch, I thought for some reason `findVirtual`'s checked exception type was `Throwable` but that is only for the `invoke` call.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:114]: I would go with `debug` as it won't affect runtime of an application, aside not having a warm startup.

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:114]: Makes sense, thanks @alzimmermsft

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/http/ReactorNettyClient.java:57]: Yes, please update this.

**Review by kushagraThapar — APPROVED** [review]
Thanks a lot @alzimmermsft  for this contribution for cosmos http client, appreciate your effort on this. 
Approving on 2 pending comments, please resolve them and we are good to merge this in afterwards.

**Pull request merged**